### ISSUE TITLE
Using torch.nn.parallel.DistributedDataParalle with DALI

### ISSUE THREAD

**Issue by dariodematties** [NVIDIA/DALI#2394]
Hi, I have an interactive pipe1 -> pipe2-> NN workflow which is explained [here](https://github.com/NVIDIA/DALI/issues/2378)
I want to parallelize this in a distributed memory system which has 2 GPUs per node
I want to put one pipe1 -> pipe2-> NN apparatus per process (rank) and map one rank per GPU.
As a result I will have two (pipe1 -> pipe2-> NN)s per node
I am using [this example](https://github.com/NVIDIA/DALI/blob/1e9196702d991d3342ad7a5a7d57c2893abad832/docs/examples/use_cases/pytorch/resnet50/main.py)
I am sending you my file so you can see everything there, but the error occurs in the following lines of code

```
210         # For distributed training, wrap the model with torch.nn.parallel.DistributedDataParallel.
211         if args.distributed:
212                 model = DDP(model, device_ids=[args.gpu], output_device=args.gpu)
213                 if args.verbose:
214                         print('Since we are in a distributed setting the model is replicated here in rank {}' .format(args.local_rank))
```

the output from the run is the following:
please ignore the three first ERRORS since they are related to containerization issues I guess

```
Singularity> python3.8 -m torch.distributed.launch --nproc_per_node=2 Contrastive_Learning.py --b 128 -v /projects/neurophon/
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
distributed is True, then rank number 1 is mapped in device number 1
Using device cuda:1 in rank number 1
distributed is True, then rank number 0 is mapped in device number 0
Using device cuda:0 in rank number 0
function_f created from rank 0
function_g created from rank 0
SimCLR_Module created from rank 0
function_f created from rank 1
function_g created from rank 1
SimCLR_Module created from rank 1
pipe1 built by rank number 0 in 3.098402500152588 seconds
pipe1 built by rank number 1 in 2.916426658630371 seconds
Initialating fixation

Initialating fixation

pipe2 built by rank number 0 in 0.0037240982055664062 seconds
pipe2 built by rank number 1 in 0.0032961368560791016 seconds
Traceback (most recent call last):
  File "Contrastive_Learning.py", line 294, in <module>
    main()
  File "Contrastive_Learning.py", line 212, in main
    model = DDP(model, device_ids=[args.gpu], output_device=args.gpu)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 335, in __init__
    self._ddp_init_helper()
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 441, in _ddp_init_helper
    self.reducer = dist.Reducer(
RuntimeError: CUDA error: the launch timed out and was terminated
terminate called after throwing an instance of 'dali::CUDAError'
  what():  CUDA runtime API error cudaErrorLaunchTimeout (6):
the launch timed out and was terminated
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/torch/distributed/launch.py", line 261, in <module>
    main()
  File "/usr/local/lib/python3.8/site-packages/torch/distributed/launch.py", line 256, in main
    raise subprocess.CalledProcessError(returncode=process.returncode,
subprocess.CalledProcessError: Command '['/usr/local/bin/python3.8', '-u', 'Contrastive_Learning.py', '--local_rank=1', '--b', '128', '-v', '/projects/neurophon/']' died with <Signals.SIGABRT: 6>.
Singularity> Traceback (most recent call last):
  File "Contrastive_Learning.py", line 294, in <module>
    main()
  File "Contrastive_Learning.py", line 212, in main
    model = DDP(model, device_ids=[args.gpu], output_device=args.gpu)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 335, in __init__
    self._ddp_init_helper()
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 441, in _ddp_init_helper
    self.reducer = dist.Reducer(
RuntimeError: CUDA error: the launch timed out and was terminated
terminate called after throwing an instance of 'dali::CUDAError'
  what():  CUDA runtime API error cudaErrorLaunchTimeout (6):
the launch timed out and was terminated
```

Basically the error is on line 212

What am I doing wrong?

Thanks! 
[Contrastive_Learning.zip](https://github.com/NVIDIA/DALI/files/5432049/Contrastive_Learning.zip)

**Comment by JanuszL** [comment]
Hi,
To be honest, I don't see anything wrong with your code, still, I'm not the most experienced PyTorch user. I don't see any `pipe.run()` invocation so I have no idea where this `terminate called after throwing an instance of 'dali::CUDAError'
  what():  CUDA runtime API error cudaErrorLaunchTimeout (6):` may came from. Have you tried to comment out the DALI pipeline and see what happens?
Can you try to trim your example and make it self contained so I can run it on our side? The zip you provided lacks many modules necessary to run it. You don't have to provide the whole code, only the stubs that would reproduce the error would do.

**Comment by dariodematties** [comment]
Thank you @JanuszL 

I commented out all DALI pipeline creations but the error persists.

DALI is not mentioned this time though

```
Singularity> python3.8 -m torch.distributed.launch --nproc_per_node=2 Contrastive_Learning.py --b 128 -v /projects/neurophon/
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
ERROR: ld.so: object '/soft/buildtools/trackdeps/${LIB}/trackdeps.so' from LD_PRELOAD cannot be preloaded: ignored.
distributed is True, then rank number 1 is mapped in device number 1
Using device cuda:1 in rank number 1
distributed is True, then rank number 0 is mapped in device number 0
Using device cuda:0 in rank number 0
function_f created from rank 0
function_g created from rank 0
SimCLR_Module created from rank 0
function_f created from rank 1
function_g created from rank 1
SimCLR_Module created from rank 1
Traceback (most recent call last):
  File "Contrastive_Learning.py", line 294, in <module>
    main()
  File "Contrastive_Learning.py", line 212, in main
    model = DDP(model, device_ids=[args.gpu], output_device=args.gpu)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 335, in __init__
    self._ddp_init_helper()
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 441, in _ddp_init_helper
    self.reducer = dist.Reducer(
RuntimeError: CUDA error: the launch timed out and was terminated
NCCL error in: /pytorch/torch/lib/c10d/../c10d/NCCLUtils.hpp:69, unhandled cuda error, NCCL version 2.4.8
Traceback (most recent call last):
  File "Contrastive_Learning.py", line 294, in <module>
    main()
  File "Contrastive_Learning.py", line 212, in main
    model = DDP(model, device_ids=[args.gpu], output_device=args.gpu)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 335, in __init__
    self._ddp_init_helper()
  File "/usr/local/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 441, in _ddp_init_helper
    self.reducer = dist.Reducer(
RuntimeError: CUDA error: the launch timed out and was terminated
NCCL error in: /pytorch/torch/lib/c10d/../c10d/NCCLUtils.hpp:69, unhandled cuda error, NCCL version 2.4.8
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/torch/distributed/launch.py", line 261, in <module>
    main()
  File "/usr/local/lib/python3.8/site-packages/torch/distributed/launch.py", line 256, in main
    raise subprocess.CalledProcessError(returncode=process.returncode,
subprocess.CalledProcessError: Command '['/usr/local/bin/python3.8', '-u', 'Contrastive_Learning.py', '--local_rank=1', '--b', '128', '-v', '/projects/neurophon/']' died with <Signals.SIGABRT: 6>.
Singularity> exit
exit
```
Any Idea of what could be happening?

I will send you a self contained zip file with the complete code

Thanks!

**Comment by JanuszL** [comment]
No idea, to be honest. Please check in https://discuss.pytorch.org.

**Comment by dariodematties** [comment]
Thank you @JanuszL I think I will better check in the Pytorch forum

Thanks!

**Comment by dariodematties** [comment]
Even though, tell me if you are interested, I can still send you the executable file. 

Thank you for your help @JanuszL !

**Comment by JanuszL** [comment]
I think you can share it, maybe it will be an inspiration for other DALI adopters in the future.
When you finish your project and you plan to publish it or make a paper let us know. We love to collect all success stories we may gather.

**Comment by dariodematties** [comment]
I will @JanuszL , in fact, if you are ok with this I will include your support in the acknowledgement section!

I am really grateful for all the support you have given to me during these weeks.

**Comment by JanuszL** [comment]
> I will @JanuszL , in fact, if you are ok with this I will include your support in the acknowledgement section!

I think the whole DALI team deserves it. I'm just its voice. 😄

**Comment by dariodematties** [comment]
Definitely! I was thinking on all the community

Thank you very much

**Comment by dariodematties** [comment]
Here is the code @JanuszL , thank you
[Contrastive_Learning.zip](https://github.com/NVIDIA/DALI/files/5439407/Contrastive_Learning.zip)
